### PR TITLE
clarify safety documentation of ptr::swap and ptr::copy

### LINKED ISSUE #81005: Safety of ptr::swap is unclear
The [`ptr::swap`](https://doc.rust-lang.org/std/ptr/fn.swap.html) function allows passed pointers to overlap, and its safety section is as follows:
> Behavior is undefined if any of the following conditions are violated:
>
>    - Both x and y must be valid for both reads and writes.
>    - Both x and y must be properly aligned.
>
> Note that even if T has size 0, the pointers must be non-NULL and properly aligned.

It says that both pointers must be "valid," but IIUC, the definition of validity becomes unclear under Stacked Borrows.

Consider the following program (all code in this comment is checked using Miri with `MIRIFLAGS="-Zmiri-track-raw-pointers"`):
```rust
unsafe {
    let mut x: [i32; 3] = [100, 200, 300];
    let p: *mut i32 = x.as_mut_ptr();
    let p1: *mut [i32; 2] = p as *mut [i32; 2]; // pointer to the first two elements
    let p2: *mut [i32; 2] = &mut *(p.add(1) as *mut [i32; 2]); // pointer to the last two elements

    // ☆: each pointer is valid according to Stacked Borrows.
    *p2 = [100, 200];
    *p1 = [200, 300];
}
```

`p1` and `p2` are both "valid" at ☆ in the sense that this program is safe.
`p2` is invalidated when we perform a write using `p1`, but it's ok because we don't use `p2` anymore.

However, the following program exhibits UB in spite of the validity of `p1` and `p2`, because `ptr::swap` first reads from `p1` and the read invalidates `p2`.
```rust
// Example 1 (UB)
unsafe {
    let mut x: [i32; 3] = [100, 200, 300];
    let p: *mut i32 = x.as_mut_ptr();
    let p1: *mut [i32; 2] = p as *mut [i32; 2];
    let p2: *mut [i32; 2] = &mut *(p.add(1) as *mut [i32; 2]); // intermediate reference

    std::ptr::swap(p1, p2); // this is UB
}
```

Note that the intermediate unique reference (`&mut *`) is crucial for "separating" the two pointers.
If we remove the intermediate reference, those pointers have the same capabilities.
In other words, the following program does **not** exhibit UB:
```rust
// Example 2 (no UB)
unsafe {
    let mut x: [i32; 3] = [100, 200, 300];
    let p: *mut i32 = x.as_mut_ptr();
    let p1: *mut [i32; 2] = p as *mut [i32; 2];
    let p2: *mut [i32; 2] = p.add(1) as *mut [i32; 2]; // no intermediate reference

    std::ptr::swap(p1, p2); // this is not UB
}
```

The difference between the two examples is so subtle that I'm not sure how to spell out the exact safety conditions for `ptr::swap`.

This issue is found while formalizing the safety of `ptr::swap` in the Rustv project.

### PR DESCRIPTION
**Pull request by RalfJung**
Closes https://github.com/rust-lang/rust/issues/81005

### CODE REVIEW

File: library/core/src/intrinsics.rs [hunk 1]
```diff
@@ -2709,6 +2709,9 @@ pub const unsafe fn copy_nonoverlapping<T>(src: *const T, dst: *mut T, count: us
 ///
 /// * `dst` must be [valid] for writes of `count * size_of::<T>()` bytes.
 ///
+/// * `src` must remain valid for reads even after `dst` is written, and vice versa.
+///   (In other words, there cannot be aliasing restrictions on the use of these pointers.)
+///
```
Comment (m-ou-se) [inline library/core/src/intrinsics.rs:2714]: I didn't quite understand what this meant until I checked out #81005.

Maybe there's a different way of wording this?

Would it work to not add this bullet point, but instead add "during the entire call to swap()" to the first two bullet points? (Maybe with a clarifying note saying that this means that the src and dst regions may not overlap.)

Comment (RalfJung) [reply library/core/src/intrinsics.rs:2714]: They may overlap though, the pointers just must allow overlapping access. So for instance passing `ptr` and `ptr.add(N)` is fine even if things overlap. However passing `&mut *ptr` twice generates two fresh mutable references and now if things overlap, that lead to UB due to aliasing violations.

It's hard to precisely say this while the aliasing model is still being determined.

> add "during the entire call to swap()" to the first two bullet points?

If we do that I think we have to say what happens during the call. So we could say something like

* `src` must be [valid] for reads of `count * size_of::<T>()` bytes, and must remain valid even if `dst` is written to for `count * size_of::<T>()` bytes.

But that seems very similar to the current wording, so I am not sure if it is clear enough.

**Comment by RalfJung** [comment]
Not sure if this wording is any better, but it's the best I could come up with so far.
@rustbot ready

**Comment by RalfJung** [comment]
@m-ou-se I think I replied to all your comments -- what's the next step for this PR?

**Review by m-ou-se — APPROVED** [review]
Looks good, thanks!

Small nit: Would it be better to say `even when` or `even while` instead of `even if`? (The `if` makes it sound a bit like those things only happen conditionally.)

r=me either way.

File: library/core/src/intrinsics.rs [hunk 2]
```diff
@@ -2705,9 +2705,13 @@ pub const unsafe fn copy_nonoverlapping<T>(src: *const T, dst: *mut T, count: us
 ///
 /// Behavior is undefined if any of the following conditions are violated:
 ///
-/// * `src` must be [valid] for reads of `count * size_of::<T>()` bytes.
+/// * `src` must be [valid] for reads of `count * size_of::<T>()` bytes, and must remain valid even
+///   if `dst` is written for `count * size_of::<T>()` bytes. (This means if the memory ranges
```
Comment (RalfJung) [inline library/core/src/intrinsics.rs:2709]:
```suggestion
///   when `dst` is written for `count * size_of::<T>()` bytes. (This means if the memory ranges
```

File: library/core/src/ptr/mod.rs [hunk 3]
```diff
@@ -795,7 +795,9 @@ pub const fn slice_from_raw_parts_mut<T>(data: *mut T, len: usize) -> *mut [T] {
 ///
 /// Behavior is undefined if any of the following conditions are violated:
 ///
-/// * Both `x` and `y` must be [valid] for both reads and writes.
+/// * Both `x` and `y` must be [valid] for both reads and writes. They must remain valid even if the
```
Comment (RalfJung) [inline library/core/src/ptr/mod.rs:798]:
```suggestion
/// * Both `x` and `y` must be [valid] for both reads and writes. They must remain valid even when the
```

**Comment by RalfJung** [comment]
> Small nit: Would it be better to say even when or even while instead of even if? (The if makes it sound a bit like those things only happen conditionally.)

My mother tongue uses the same word for "if" and "when" so this is probably just me missing a connotation that will be obvious to native speakers. I changed the wording.

@bors r=m-ou-se

**Pull request merged**